README.md
## [中文](README-ZH.md)

# racoon-js
The easiest to use and powerful data validate and formatter library for JS.

[![npm version](https://img.shields.io/npm/v/racoon-js.svg?style=flat-square)](https://www.npmjs.com/package/racoon-js)
[![build status](https://img.shields.io/travis/charlyzeng/racoon-js/master.svg?style=flat-square)](https://travis-ci.org/charlyzeng/racoon-js)
[![code coverage](https://img.shields.io/coveralls/charlyzeng/racoon-js.svg?style=flat-square)](https://coveralls.io/r/charlyzeng/racoon-js)
[![install size](https://packagephobia.com/badge?p=racoon-js)](https://packagephobia.com/result?p=racoon-js)
[![npm downloads](https://img.shields.io/npm/dm/racoon-js.svg?style=flat-square)](http://npm-stat.com/charts.html?package=racoon-js)

## Table Of Contents
- [API Document](/doc/API.md)
- [Features](#features)
- [Browser Support](#browser-support)
- [Installing](#installing)
- [Example](#example)

## Features
- 🌈 Support both browser and node.
- ❄️ The data format can be flexibly defined, and any level of attribute nesting of object and array is supported.
- 🔗 Support chain call without paying attention to call sequence.
- ✨ Error message can be easily customed.

## Browser Support
![Chrome](https://raw.github.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png) | ![Firefox](https://raw.github.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png) | ![Safari](https://raw.github.com/alrra/browser-logos/master/src/safari/safari_48x48.png) | ![Opera](https://raw.github.com/alrra/browser-logos/master/src/opera/opera_48x48.png) | ![Edge](https://raw.github.com/alrra/browser-logos/master/src/edge/edge_48x48.png) | ![IE](https://raw.github.com/alrra/browser-logos/master/src/archive/internet-explorer_9-11/internet-explorer_9-11_48x48.png) |
--- | --- | --- | --- | --- | --- |
Latest ✔ | Latest ✔ | Latest ✔ | Latest ✔ | Latest ✔ | 11 ✔ |

## Installing
Using npm:
```bash
$ npm install racoon-js
```

Using yarn:
```bash
yarn add racoon-js
```

Using cdn:
```html
<script src="https://cdn.jsdelivr.net/npm/racoon-js@latest/dist/racoon.min.js"></script>
```
The JS from cdn is _21kB_, and it will be only _4.6kB_ after be compressed by gzip.

## Example
```javascript
import racoon from 'racoon-js';

// Define data format schema
const schema = racoon.object({
  name: racoon
    .string()
    .min(3)
    .error('name is too short') // You can add your custom error message
    .max(30)
    .error('name is too long')
    .required(), // Name can not be `undefined` or `null`
  age: racoon
    .number()
    .int()
    .default(1) // If value is `null` or `undefined`, then return default value `1`
  married: racoon.boolean(),
  favorite: racoon.object({ // You can define deep nested object format schema
    sports: racoon.array(
      racoon.string().min(1).max(100).required()
    ),
    book: racoon.object({
      title: racoon.string().required(),
      date: racoon
        .string()
        .pattern(/^\d{4}-\d{1,2}-\d{1,2}$/)
        .format((date) => { // Format the return value
          return `${date} 12:00:00`;
        })
    })
  })
});

try {
  const result = schema.validate({
    name: 'Jack',
    age: 22,
    favorite: {
      sports: ['football', 'basketball'],
      book: {
        title: 12,
        date: '2010-1-1'
      }
    }
  });
} catch (error) {
  console.error(error);
  // Error: "favorite.book.title": value should be a type of string
  // ↑↑↑ The error message can be customed
}

// If you don't like `try-catch` code style,
// you can use `validateSilent` method. When
// validate failed, the `error` will not be
// empty. Otherwise，the `error` is `undefined`.
//
// For Example:
const { error, value } = schema.validateSilent({ ... });
```


CHANGELOG.md
# Changelog

### 0.0.1 (Aug 2, 2020)
- First beta version

### 1.0.0 (Aug 16, 2020)
- Initial stable release

### 1.1.0 (Aug 17, 2020)
- Add `index.d.ts`

### 1.2.0 (Aug 18, 2020)
- Add source map

### 1.2.1 (Aug 19, 2020)
- Remove `Symbol` from code
- Remove some useless code

### 1.2.2 (Aug 21, 2020)
- Fix problem that custom error message contains key chain

### 1.3.0 (Aug 23, 2020)
- Add `allowString` method for number type
- Add `errorForAll` method for all type
- The callback arg of `custom` method support to return a string to respect failure reason

### 1.3.1 (Aug 24, 2020)
- Update `index.d.ts` to fix grammar prompt

### 1.3.2 (Aug 25, 2020)
- Fix the problem that `allowString` accept empty-string and the objects which `toString` return a number style string

### 1.3.3 (Aug 27, 2020)
- Fix bug about `object` and `array`

### 1.3.4 (Aug 27, 2020)
- Fix bug when use `stripUnkown`

### 1.3.5 (Sep 2, 2020)
- Update README and keywords

### 1.3.6 (Oct 7, 2020)
- Fix bug of `default not effect` when key not exist in object.

### 1.3.7 (Oct 7, 2020)
- Fix bug of `required not effect` when key not exist in object.

### 1.3.8 (Dec 19, 2020)
- Modify `index.d.ts`.

### 1.3.9 (Dec 20, 2020)
- Modify `index.d.ts`.

### 1.4.0 (Jan 14, 2021)
- Support `mixed` schema.

### 1.5.0 (Jan 16, 2021)
- Support `even` and `odd` for number schema.

### 1.6.0 (Jan 17, 2021)
- Reduce the size of bundle.


karma.conf.js
// Karma configuration
// Generated on Tue Jan 05 2021 17:54:02 GMT+0800 (GMT+08:00)

const resolve = require('@rollup/plugin-node-resolve').default;
const babel = require('@rollup/plugin-babel').default;
const commonjs = require('@rollup/plugin-commonjs');
const json = require('@rollup/plugin-json');

const watch = Object.prototype.hasOwnProperty.call(process.env, 'WATCH');

module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'detectBrowsers'],


    // list of files / patterns to load in the browser
    files: [
      'test/**/*.test.js',
    ],


    // list of files / patterns to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*.test.js': ['rollup'],
    },

    rollupPreprocessor: {
      output: {
        format: 'umd',
        name: 'racoon',
        sourcemap: 'inline',
      },
      plugins: [
        json(),
        commonjs(),
        resolve(),
        babel({
          exclude: 'node_modules/**',
          babelHelpers: 'runtime',
        }),
      ],
    },

    detectBrowsers: {
      // enable/disable, default is true
      enable: true,

      // enable/disable phantomjs support, default is true
      usePhantomJS: false,

      // use headless mode, for browsers that support it, default is false
      preferHeadless: false,

      // post processing of browsers list
      // here you can edit the list of browsers used by karma
      postDetection(installedBrowsers) {
        // Can not launch Edge. For detail to see the follow issue:
        // https://github.com/MicrosoftEdge/edge-launcher/issues/23
        return installedBrowsers.filter(browser => browser !== 'Edge');
      },
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: watch,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    // browsers: ['Chrome', 'Firefox', 'Safari', 'Edge'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: !watch,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity,
  });
};
